prow/cmd/peribolos/main.go

```diff
@@ -0,0 +1,137 @@
+
+package main
+
+import (
+	"errors"
+	"flag"
+	"fmt"
+	"io/ioutil"
+	"log"
+	"net/url"
+	"os"
+	"strings"
+
+	"k8s.io/test-infra/prow/config"
+	"k8s.io/test-infra/prow/config/org"
+	"k8s.io/test-infra/prow/github"
+)
+
+const defaultEndpoint = "https://api.github.com"
+
+type options struct {
+	config   string
+	token    string
+	confirm  bool
+	endpoint string
+}
+
+func parseOptions() options {
+	var o options
+	if err := o.parseArgs(flag.CommandLine, os.Args[1:]); err != nil {
+		log.Fatalf("Invalid flags: %v", err)
+	}
+	return o
+}
+
+func (o *options) parseArgs(flags *flag.FlagSet, args []string) error {
+	flags.StringVar(&o.config, "config-path", "", "Path to prow config.yaml")
+	flags.BoolVar(&o.confirm, "confirm", false, "Mutate github if set")
+	flags.StringVar(&o.endpoint, "github-endpoint", defaultEndpoint, "Github api endpoint, may differ for enterprise")
+	flags.StringVar(&o.token, "github-token-path", "", "Path to github token")
+	if err := flags.Parse(args); err != nil {
+		return err
+	}
+	if o.config == "" {
+		return errors.New("empty --config")
+	}
+
+	if o.token == "" {
+		return errors.New("empty --github-token-path")
+	}
+
+	if _, err := url.Parse(o.endpoint); err != nil {
+		return fmt.Errorf("invalid --endpoint URL: %v", err)
+	}
+	return nil
+}
+
+func main() {
+	o := parseOptions()
+
+	if o.config != "TODO(fejta): implement me" {
+		log.Fatalf("This program is not yet implemented")
+	}
+
+	cfg, err := config.Load(o.config)
+	if err != nil {
+		log.Fatalf("Failed to load --config=%s: %v", o.config, err)
+	}
+
+	b, err := ioutil.ReadFile(o.token)
+	if err != nil {
+		log.Fatalf("cannot read --token: %v", err)
+	}
+
+	var c *github.Client
+	tok := strings.TrimSpace(string(b))
+	if o.confirm {
+		c = github.NewClient(tok, o.endpoint)
+	} else {
+		c = github.NewDryRunClient(tok, o.endpoint)
+	}
+	c.Throttle(300, 100) // 300 hourly tokens, bursts of 100
+
+	for name, orgcfg := range cfg.Orgs {
+		if err := configureOrg(c, name, orgcfg); err != nil {
+			log.Fatalf("Configuration failed: %v", err)
+		}
+	}
+}
+
+func configureOrg(client *github.Client, orgName string, orgConfig org.Org) error {
+	// get meta
+	// diff meta
+	// patch meta
+
+	// people = set(all org members)
+	// remove = people - orgConfig.members - orgConfig.admins
+	// update = oc.members - [p for p in people if member]
+	// update = oc.admins - [p for p in people if admin]
+
+	// teams = set(all teams)
+	for name, team := range orgConfig.Teams {
+		team.Name = name
+		num := -1
+		// find team name in teams (or previous name)
+		if err := configureTeam(client, num, team); err != nil {
+			return fmt.Errorf("failed to update %s: %v", orgName, err)
+		}
+	}
+	return nil
+}
+
+func configureTeam(client *github.Client, num int, team org.Team) error {
+	// create team if num < 0
+	// else diff and patch
+
+	// people = set(all org members)
+	// remove = people - orgConfig.members - orgConfig.admins
+	// update = oc.members - [p for p in people if member]
```

Comment:
@fejta how can we tell you think in Python? :stuck_out_tongue:

Reply:
:heart: list comprehensions :heart: